Assets/Mirror/Runtime/NetworkConnectionToClient.cs

```diff
@@ -9,16 +10,124 @@ public class NetworkConnectionToClient : NetworkConnection
+            // -> initialize with current NetworkTime so first update doesn't
+            //    calculate elapsed via 'now - 0'
+            internal double lastSendTime = NetworkTime.time;
+        }
+        Dictionary<int, Batch> batches = new Dictionary<int, Batch>();
+
+        // batch interval is 0 by default, meaning that we send immediately.
+        // (useful to run tests without waiting for intervals too)
+        float batchInterval;
+
+        public NetworkConnectionToClient(int networkConnectionId, float batchInterval)
+            : base(networkConnectionId)
+        {
+            this.batchInterval = batchInterval;
+        }
+
+        Batch GetBatchForChannelId(int channelId)
+        {
+            // get existing or create new writer for the channelId
+            Batch batch;
+            if (!batches.TryGetValue(channelId, out batch))
+            {
+                batch = new Batch();
+                batches[channelId] = batch;
+            }
+            return batch;
+        }
+
+        void SendBatch(int channelId, Batch batch)
+        {
+            // send batch
+            Transport.activeTransport.ServerSend(connectionId, channelId, batch.writer.ToArraySegment());
+
+            // clear batch
+            batch.writer.Reset();
+
+            // reset send time for this channel's batch
+            batch.lastSendTime = NetworkTime.time;
+        }
 
         internal override void Send(ArraySegment<byte> segment, int channelId = Channels.DefaultReliable)
         {
             if (logger.LogEnabled()) logger.Log("ConnectionSend " + this + " bytes:" + BitConverter.ToString(segment.Array, segment.Offset, segment.Count));
 
+            //Debug.Log("ConnectionSend " + this + " bytes:" + BitConverter.ToString(segment.Array, segment.Offset, segment.Count));
+
             // validate packet size first.
             if (ValidatePacketSize(segment, channelId))
             {
-                Transport.activeTransport.ServerSend(connectionId, channelId, segment);
+                // always batch!
+                // (even if interval == 0, in which case we flush in Update())
+                //
+                // if batch would become bigger than MaxBatchPacketSize for this
+                // channel then send out the previous batch first.
+                //
+                // IMPORTANT: we use maxBATCHsize not maxPACKETsize.
+                //            some transports like kcp have large maxPACKETsize
+                //            like 144kb, but those would extremely slow to use
+                //            all the time for batching.
+                //            (maxPACKETsize messages still work fine, we just
+                //             aim for maxBATCHsize where possible)
+                Batch batch = GetBatchForChannelId(channelId);
+                int max = Transport.activeTransport.GetMaxBatchSize(channelId);
+                if (batch.writer.Position + segment.Count > max)
```

Comment:
```suggestion
                if (batch.writer.Length+ segment.Count > max)
```

This should be written length, not position.